pkg/webhooks/kafkacluster_validator_test.go

```diff
@@ -486,3 +487,221 @@ func TestCheckBrokerStorageRemoval(t *testing.T) {
 		}
 	}
 }
+
+func TestCheckExternalListenerStartingPort(t *testing.T) {
```

Comment:
Question about the combination of broker ID and `ExternalStartingPort` used in the test: is it possible to use some randomly generated numbers that meet the expected behavior? 

The main reason behind this ask is that the current implementation always uses the fixed combination of broker IDs and `ExternalStartingPort` no matter how many times we run the unit test, which is not very ideal from the testing perspective. The best scenario is to use some randomly generated numbers within certain range every time that the test is run.

Reply:
> current implementation always uses the fixed combination of broker IDs and ExternalStartingPort no matter how many times we run the unit test

The test does that because that's what we do in the operator code.

> is it possible to use some randomly generated numbers that meet the expected behavior?

A longwinded reply about random numbers for those 2 user inputs (broker ids and startingExternalPort) : 
1. I don't think that's attainable because the main condition we ultimately check against in this function is `RNG_1 + RNG_2 <= 65535` which translates to nothing meaningful in limiting the range of the 2 numbers other than that both have to be < 65535. 
    * We do that in this PR via the OpenAPIv3Schema validations (markers) but because that is such a weak condition (means close to nothing, in practice) we add the validating webhook function to check the sum for more practical relevance.
    * How would you generate numbers that don't "obviously/by default" satisfy these conditions **AND know in advance** that the case should result in true/false (aka numbers and the sum are valid/invalid) ?
2. I believe the current setup proves considerably more even as documentation value because the current 3 test cases within that function translate in natural language to : 
    * [1] obviously ok / well in range
    * [2] obviously bad / well out of range
    * [3] borderline values - some results are in range, some are out of range but sometimes barely out of range which emphasises the limits of the intended interval.     

I believe addressing these cases is also part of general testing theory.

Due to the lack of a meaningful mathematical condition that we could enforce **a-priori**, I think random numbers could only replace the ones in the "obviously ok/bad" cases (never the borderline cases) and that, in my opinion, doesn't add value to our tests.   
Yes, in the "obvious" cases, I pulled those broker ids and startingPorts out of "thin air" but making them smth like (for the "obviously ok" case):
```
brokers[0-2].id = rng(0<x<999)
listenersConfig.[...int/ext...].externalStartingPort = rng(15000 < x < 50000)
```
is, imo, just a headache without a payoff.

Does that thought process make sense ?

Reply:
I disagree that using random numbers doesn't add value to the tests, I am a bit lazy of writing a lot of text so let's look at the code...

```golang

	const lowerEnd int32 = 30000

	testCases := []struct {
		testName         string
		kafkaClusterSpec v1beta1.KafkaClusterSpec
		expected         field.ErrorList
	}{
		{
			// In this test case, all resulting external port numbers should be valid
			testName: "valid config: 3 brokers with 2 externalListeners",
			kafkaClusterSpec: v1beta1.KafkaClusterSpec{
				Brokers: []v1beta1.Broker{{Id: rand.Int31n(lowerEnd)}, {Id: rand.Int31n(lowerEnd)}, {Id: rand.Int31n(lowerEnd)}},
				ListenersConfig: v1beta1.ListenersConfig{
					ExternalListeners: []v1beta1.ExternalListenerConfig{
						{
							CommonListenerSpec:   v1beta1.CommonListenerSpec{Name: "test-external1"},
							ExternalStartingPort: rand.Int31n(65535 - lowerEnd),
						},
						{
							CommonListenerSpec:   v1beta1.CommonListenerSpec{Name: "test-external2"},
							ExternalStartingPort: rand.Int31n(65535 - lowerEnd),
						},
					},
				},
			},
			expected: nil,
		},
...

```

This way we make sure that the `checkExternalListenerStartingPort` actually behaves as we expected with different combination of numbers with the broker.id and the starting port number every time that we run the unit test, which I think it's much better

Reply:
I think I see what you mean. I like it for this case (the previously deemed "obviously ok" case).  
Do you see it applied on other cases ?

Reply:
> I think I see what you mean. I like it for this case (the previously deemed "obviously ok" case aka when `expected` is always `nil`). Do you see it applied on other cases ?

yea, I think we can apply the same logic to the other two test cases under this test function by setting the range for the randomly generated numbers

Reply:
While Darren's proposal does have it's merit, it should be done during fuzzy testing IMO instead of unit testing.
In unit tests I would rather adhere to the TDD FIRST principle: https://stackoverflow.com/questions/18024785/tdd-first-principle in which the repeatability/reproducibility would be much better with non-random input data even if semantically the results should be the same. (The reason is unexpected breakings depending on the random input.)

Reply:
> I think we can apply the same logic to the other two test cases under this test function by setting the range for the randomly generated numbers

I was looking today at it and one challenge I see for the cases when `expected` is not `nil` (especially in the context of [this thread](https://github.com/banzaicloud/koperator/pull/939#discussion_r1119122154) and us going the "exact match" route) is defining the "precise" `expected` results because they would need to include most of these random numbers and the name of the listeners which fail.

For example (snippet partially defining `expected` from another test case ):  
```
expected: append(field.ErrorList{},
	field.Invalid(field.NewPath("spec").Child("listenersConfig").Child("externalListeners").Index(0).Child("externalStartingPort"), int32(65535),
		invalidExternalListenerStartingPortErrMsg+": "+fmt.Sprintf("ExternalListener '%s' would generate invalid port numbers (not between 1 and 65535) for brokers %v",
			"test-external1", []int32{11, 102})),
... ... ...
```

In that example, I don't know how to apply the random numbers idea to replace : 
* `int32(65535)`
    * this is the `value interface{}` parameter of the `field.Invalid()` function
    * this challenge applies to generating any `field.Error` aka as long as `expected` is not `nil`
 
Similar idea applies to these but they are not unavoidable in the webhook function: 
* `[]int32{11, 102}`
    * this is a bit of helping info about which broker ids are incompatible with the externalListener 
* `"test-external1"`
    * least important, just makes the log a bit nicer in my opinion but I do like having it in the log message

Reply:
Thanks @pregnor for the pointer! I wasn't aware the TDD FIRST principle in testing due to experience. Based on that, I think you can ignore my original comment now @mihaialexandrescu